arith/mersenne31/src/m31_ext6.rs

```diff
@@ -0,0 +1,327 @@
+}
+
+field_common!(M31Ext6);
+
+impl ExpSerde for M31Ext6 {
+    const SERIALIZED_SIZE: usize = (32 / 8) * 6;
+
+    #[inline(always)]
+    fn serialize_into<W: Write>(&self, mut writer: W) -> SerdeResult<()> {
+        self.v[0].serialize_into(&mut writer)?;
+        self.v[1].serialize_into(&mut writer)?;
+        Ok(())
+    }
+
+    // FIXME: this deserialization function auto corrects invalid inputs.
+    // We should use separate APIs for this and for the actual deserialization.
+    // NOTE(HS): directly copied from M31ext3 impl
+    #[inline(always)]
+    fn deserialize_from<R: Read>(mut reader: R) -> SerdeResult<Self> {
+        Ok(Self {
+            v: [
+                M31Ext3::deserialize_from(&mut reader)?,
+                M31Ext3::deserialize_from(&mut reader)?,
+            ],
+        })
+    }
+}
+
+impl Field for M31Ext6 {
+    const NAME: &'static str = "Mersenne 31 Extension 6";
+
+    const SIZE: usize = 32 / 8 * 6;
+
+    const FIELD_SIZE: usize = 32 * 6;
+
+    const ZERO: Self = M31Ext6 {
+        v: [M31Ext3::ZERO, M31Ext3::ZERO],
+    };
+
+    const ONE: Self = M31Ext6 {
+        v: [M31Ext3::ONE, M31Ext3::ZERO],
+    };
+
+    const INV_2: M31Ext6 = M31Ext6 {
+        v: [M31Ext3::INV_2, M31Ext3::ZERO],
+    };
+
+    const MODULUS: U256 = M31::MODULUS;
+
+    #[inline(always)]
+    fn zero() -> Self {
+        Self::ZERO
+    }
+
+    #[inline(always)]
+    fn is_zero(&self) -> bool {
+        self.v[0].is_zero() && self.v[1].is_zero()
+    }
+
+    #[inline(always)]
+    fn one() -> Self {
+        Self::ONE
+    }
+
+    fn random_unsafe(mut rng: impl RngCore) -> Self {
+        Self {
+            v: [
+                M31Ext3::random_unsafe(&mut rng),
+                M31Ext3::random_unsafe(&mut rng),
+            ],
+        }
+    }
+
+    fn random_bool(mut rng: impl RngCore) -> Self {
+        M31Ext6 {
+            v: [M31Ext3::random_bool(&mut rng), M31Ext3::ZERO],
+        }
+    }
+
+    fn inv(&self) -> Option<Self> {
+        if self.is_zero() {
+            return None;
+        }
+
+        let normalize = (-self.v[0].square() - self.v[1].square().double()).inv()?;
+
+        let compliment = Self {
+            v: [-self.v[0] * normalize, self.v[1] * normalize],
+        };
+
+        Some(compliment)
+    }
+
+    /// Squaring
+    #[inline(always)]
+    fn square(&self) -> Self {
+        Self {
+            v: square_internal(&self.v),
+        }
+    }
+
+    #[inline(always)]
+    fn as_u32_unchecked(&self) -> u32 {
+        self.v[0].as_u32_unchecked()
+    }
+
+    #[inline(always)]
+    fn from_uniform_bytes(bytes: &[u8; 32]) -> Self {
+        let v1 = mod_reduce_u32(u32::from_be_bytes(bytes[0..4].try_into().unwrap()));
+        let v2 = mod_reduce_u32(u32::from_be_bytes(bytes[4..8].try_into().unwrap()));
+        let v3 = mod_reduce_u32(u32::from_be_bytes(bytes[8..12].try_into().unwrap()));
+
+        // TODO(HS) need some inputs - do we need extra mod reduce u32 like m31ext3 impl?
```

Comment:
there is a small chance that for uniform u32 inputs, a single mod reduce will produce 2^31 or 2^31-1 so i think we should take two mod reduction here

Reply:
im a bit unsure so I am unresolving this one - seems `mod_reduce_u32` reduces inputs into range `[0, p]`, but rather we want something like `[0, p - 1]`, an additional call to `mod_reduce_u32` won't help

asking for suggestions here - do we want to fix the behavior that m31 from random bytes to strict `[0,p-1]` range? otherwise, we can just remove the additional call to mod reduce u32, as that won't help...

Reply:
see https://github.com/PolyhedraZK/Expander/pull/257/commits/ac7889a83a66fe927461f14d5b27ece06449c91a per offline discussion.